design/baremetal-operator/image-builder-integration.md

```diff
@@ -0,0 +1,241 @@
+- Add a watch on the `RamDiskImageRequest` for the `BareMetalHost` reconciler
+- Create the `RamDiskImageRequest` when it is missing
+- Pause registration until the image URL is available
+- Pass the ramdisk URL to Ironic
+
+### Dependencies
+
+None
+
+### Test Plan
+
+Enable the functionality in the integration test, using the default (trivial)
+controller implementation for the `RamDiskImageRequest`. This should produce
+the same result as today, but exercise the integration between the the two
+resource types.
+
+### Upgrade / Downgrade Strategy
+
+This functionality will be disabled by default, so that existing users don't
+need to do anything special on upgrade.
+
+### Version Skew Strategy
+
+The API surface area is very small, so we are unlikely to encounter version
+skew issues.
+
+## Drawbacks
+
+This adds complexity inside the baremetal-operator, for something that is not
+needed by many users.
+
+## Alternatives
+
+### Have Ironic customise the image
+
+Currently Ironic just boots the Node with a given kernel and initrd. If Ironic
+were able to supply the networkData to the Node as well when booting the
+ramdisk image, then customisation could be performed within the ramdisk. It
+appears that theoretically Ironic can already [pass the networkData in this
+way](https://docs.openstack.org/ironic/latest/admin/dhcp-less.html#configuring-network-data)
+when creating an ISO to boot the deploy image using virtualmedia. This should
+be enough to accomplish our goals here. It is implemented in the iLO and
+redfish drivers, and inherited by idrac-redfish-virtualmedia.
+
+One downside of this is that there is no access from outside for a user to
+request their own image build (e.g. to manually boot a server into the deploy
+ramdisk for discovery).
+
+With this strategy we would always have to update the network data when it
+changes; we have no way of distinguishing between changes made by an
+administrator or other user.
+
+### Configure the IP address from within a fixed image
+
+Using IPv6 Neighbour Discovery, we can determine the network prefix for each
+interface. Using the network prefix we can statelessly choose a static IP
+address without fear of collison (by using the Modified EUI-64 method based on
+the MAC address, and/or a stable privacy address).
+
+If the host is not directly attached to the network Ironic is on, Neighbour
+Discovery can also locate routers on the attached networks and identify which
+are suitable to be used as a default route.
+
+Information about VLANs and static bonds can be obtained from LLDP, and
+information about dynamic bonds from LACP or LLDP. This information could be
+used to automatically configure the interfaces on the ramdisk.
+
+This only works for IPv6, but it would be pretty unusual for an IPv4-only
+environment not to have DHCP. It also assumes that LLDP is enabled on the local
+switch, although that is common (and also relied upon by our hardware details
+introspection).
```

Comment:
This wouldn't give us the ability to use bonded NICs, right?

Reply:
It would work fine if the switch is configured with LACP (which is apparently the common case). It wouldn't work reliably if the switch is configured with a static link aggregation.